sdk/storage/azure-storage-blob/src/test/java/com/azure/storage/blob/LargeBlobTest.groovy

```diff
@@ -232,20 +320,57 @@ class LargeBlobTest extends APISpec {
             def request = httpPipelineCallContext.httpRequest
             // Substitute large body for put block requests and collect size of original body
-            if (request.url.getQuery() != null && request.url.getQuery().endsWith("comp=block")) {
-                if (collectSize) {
-                    def bytesReceived = request.getBody().reduce(0L, new BiFunction<Long, ByteBuffer, Long>() {
-                        @Override
-                        Long apply(Long a, ByteBuffer byteBuffer) {
-                            return a + byteBuffer.remaining()
-                        }
-                    })
-                    putBlockPayloadSizes.add(bytesReceived)
+            def urlString = request.getUrl().toString()
+            if (isPutBlockRequest(request)) {
+                if (!retryTracker.get(urlString, false)) {
+                    blocksCount.incrementAndGet()
+                    retryTracker.put(urlString, true)
+                }
+                Mono<Long> count = interceptBody(request)
+                if (count != null) {
+                    return count.flatMap { bytes ->
+                        putBlockPayloadSizes.add(bytes)
+                        return httpPipelineNextPolicy.process()
+                    }
+                }
+            } else if (isSinglePutBlobRequest(request)) {
+                if (!retryTracker.get(urlString, false)) {
+                    singleUploadCount.incrementAndGet()
+                    retryTracker.put(urlString, true)
+                }
+                Mono<Long> count = interceptBody(request)
+                if (count != null) {
+                    return count.flatMap { bytes ->
+                        putBlobPayloadSizes.add(bytes)
+                        return httpPipelineNextPolicy.process()
+                    }
                 }
-                count.incrementAndGet()
-                request.setBody("dummyBody")
             }
             return httpPipelineNextPolicy.process()
         }
+
+        private Mono<Long> interceptBody(HttpRequest request) {
+            Mono<Long> result = null
+            if (collectSize) {
+                result = request.getBody().reduce(0L, new BiFunction<Long, ByteBuffer, Long>() {
+                    @Override
+                    Long apply(Long a, ByteBuffer byteBuffer) {
+                        return a + byteBuffer.remaining()
+                    }
+                })
+            }
+            request.setBody("dummyBody")
+            return result
+        }
+
+        private boolean isPutBlockRequest(HttpRequest request) {
+            return request.url.getQuery() != null && request.url.getQuery().endsWith("comp=block");
+        }
+
+        private boolean isSinglePutBlobRequest(HttpRequest request) {
+            return request.getHttpMethod().equals(HttpMethod.PUT) &&
+                request.getUrl().toString().endsWith(blobName) &&
+                request.getUrl().getQuery() == null
```

Comment:
Would it be better to check for certain parts of the query string being null in case this gets used by a client authorizing with a SAS token.

Reply:
I agree. However, I don't expect adding SAS tokens scenarios here (orthogonal feature). This Policy is not intended to be used outside of large payload testing. Therefore, I'd rather keep it simple until we need more functionality here.

Reply:
Sounds good